### ISSUE TITLE
@discordjs/voice: 34% silent packet loss during DAVE MLS key transitions

### ISSUE THREAD

**Issue by wave-workflow** [discordjs/discord.js#11441]
## Which package is this bug report for?

voice (`@discordjs/voice` 0.19.0)

## Issue description

During DAVE E2E encryption, 34% of Opus voice packets are silently dropped during MLS key transitions. No errors are surfaced to the consumer — packets simply vanish. This makes voice receive unreliable for any application that processes speech (STT, recording, moderation).

Separated from #11419 per @Qjuh's request — this is a distinct issue from the reconnect-loop behavior reported there.

## Dependency Report

```
Core Dependencies
- @discordjs/voice: 0.19.0
- prism-media: 1.3.5

Opus Libraries
- @discordjs/opus: 0.10.0

Encryption Libraries
- native crypto support for aes-256-gcm: yes
- sodium-native: 5.0.10

DAVE Libraries
- @snazzah/davey: 0.1.10

FFmpeg
- version: 6.1.1-3ubuntu5
- libopus: yes
```

## Environment

- Node.js: v22.22.0
- OS: Linux 6.6.87.2-microsoft-standard-WSL2 (x64)
- Discord server: DAVE enabled (default)

## Steps to reproduce

1. Join a DAVE-enabled voice channel (now the default)
2. Subscribe to a user's audio stream via `VoiceReceiver`
3. Have the user speak continuously for 30+ seconds
4. Measure total bytes received vs. expected at the Opus bitrate

```js
const connection = joinVoiceChannel({
  channelId: channel.id,
  guildId: channel.guild.id,
  adapterCreator: channel.guild.voiceAdapterCreator,
  selfDeaf: false,
  debug: true,
});

connection.on('debug', (msg) => console.log(`[DEBUG] ${msg}`));

connection.receiver.speaking.on('start', (userId) => {
  const stream = connection.receiver.subscribe(userId, {
    end: { behavior: EndBehaviorType.AfterSilence, duration: 1000 },
  });
  let bytes = 0, chunks = 0;
  stream.on('data', (chunk) => { bytes += chunk.length; chunks++; });
  stream.on('end', () => console.log(`${bytes} bytes in ${chunks} chunks`));
});
```

## Expected behavior

All Opus packets that pass DAVE decryption should be delivered to the consumer stream. Packets that fail decryption should emit an error or be accounted for in a metric.

## Actual behavior

~34% of packets are silently dropped. No `error` event on the stream, no decryption failure in debug output. The data simply never arrives.

### Quantitative analysis

Over **55 seconds of continuous speech**, 25 audio segments captured:

| Metric | Value |
|--------|-------|
| Total bytes received | 272,894 |
| Expected bytes (64kbps Opus) | ~412,500 |
| Packet delivery rate | **66%** |
| Silent loss | **34%** |

### Segment size distribution

```
43,366 bytes (~5.8s)  ← largest, stable key period
24,077 bytes (~3.2s)
22,881 bytes (~3.1s)
 ...
   479 bytes (~0.1s)  ← near-total loss
   466 bytes (~0.1s)
   387 bytes (~0.1s)  ← smallest, during key transition
```

### Correlation with MLS key transitions

Data loss spikes during DAVE MLS operations visible in debug output:

```
[DEBUG] [NW] [DAVE] MLS commit processed (transition id: 0)
[DEBUG] [NW] [DAVE] MLS proposals processed
```

The largest segments (5+ seconds of clean audio) correspond to periods **between** key rotations. The smallest segments (<0.1s) occur **during** MLS commit/proposal processing, suggesting that packets arriving during key transitions are decrypted with a stale key and silently dropped rather than queued or retried.

## Impact

When partial audio reaches STT engines (tested: Whisper large-v3 CUDA, Gemini 2.5 Flash), the 34% gaps cause **hallucinated transcriptions** — models generate entire sentences from training data (lectures, phrases in random languages) because corrupted/missing frames decode to noise rather than silence.

This makes voice receive unusable for production bots that need reliable speech-to-text.

## Comparison with py-cord

py-cord 2.6.1 handles the same voice channel with DAVE correctly — **100% packet delivery**, zero data loss. This confirms the issue is in `@discordjs/voice`'s DAVE packet handling, not Discord's protocol or the network path.

## Possible root cause

Hypothesis: during MLS key transitions, there's a window where incoming packets are encrypted with the new key but the receiver is still using the old key (or vice versa). The decryption fails silently and the packet is dropped without surfacing an error. A brief overlap window where both keys are tried (or a small packet buffer across transitions) might resolve this.

## Suggested fix directions

1. **Dual-key window**: Keep both old and new keys active for a short overlap period during MLS transitions. Try decryption with both keys before dropping.
2. **Packet buffering**: Buffer incoming packets during key transitions (~100-200ms) and retry decryption once the new key is active.
3. **Error surfacing**: At minimum, emit a counter or event when packets fail decryption so consumers can detect data loss.

Any of these would significantly improve reliability. Option 3 is the lowest effort and would at least make the problem observable.

**Comment by Snazzah** [comment]
py-cord doesn't even support DAVE yet, let alone voice recieve: https://github.com/Pycord-Development/pycord/pull/2873

**Comment by wave-workflow** [comment]
Thanks for the correction @Snazzah — you're right. I referenced the py-cord comparison from the original issue (#11419) without verifying their DAVE support status. That comparison is invalid and I've struck it from the issue description.

The core finding stands independently of any cross-library comparison:

**Over 55 seconds of continuous speech with `@discordjs/voice` 0.19.0 + `@snazzah/davey` 0.1.10:**
- 272,894 bytes captured out of ~412,500 expected (64kbps Opus)
- **34% of packets silently dropped** — no errors on the stream, no decryption failures in debug output
- Smallest segments (~387 bytes / 0.1s) correlate with MLS commit/proposal processing in debug logs
- Largest segments (~43,366 bytes / 5.8s) occur between key rotations

The silent nature of the loss is the main concern — there's no way for the consumer to detect that data is missing. Would a packet-level trace (RTP sequence numbers before/after DAVE decryption) be helpful for debugging?

**Comment by wave-workflow** [comment]
Following up with a deeper analysis of the captured PCM data.

## Segment size distribution is bimodal

25 segments over 74.1 seconds show two distinct populations:

| Group | Count | Mean size | % of segments |
|-------|-------|-----------|---------------|
| Tiny (<2KB) | 7 | 772 bytes | 28% |
| Normal (≥2KB) | 18 | 14,860 bytes | 72% |

That's a **19× difference** between groups. Coefficient of variation is 98% — far outside what's expected for continuous speech (typically 10–20%).

## Three temporal phases are visible

```
Phase 1 (0–12s):  avg 10,397 bytes — initial connection, mixed sizes
Phase 2 (12–41s): avg  6,672 bytes — degraded, 5 of 7 tiny segments cluster here
Phase 3 (41–74s): avg 17,140 bytes — stabilized, 2.6× better than Phase 2
```

Visual (segment sizes over time, each █ ≈ 1KB):

```
 0.0s | ████████████           Phase 1
 0.7s | ████████████
 5.2s | ████
 8.5s | █████████████████
10.6s | ██████████
12.4s | █████
15.7s |                        Phase 2 — tiny segments cluster
17.0s | ████████████████████████
18.8s | ██████████████████
20.0s | ███████████████
24.6s |                        ⬅ tiny
30.7s | ██
32.6s | ███████
36.0s |                        ⬅ tiny
38.9s | █
39.8s | █
40.6s |                        ⬅ tiny
42.0s | ███████████████████████████████████████████  Phase 3
46.2s | █████████████████████████████
50.6s | ██████████████
55.0s | ██████████████████████
62.4s | ████
64.8s |                        ⬅ tiny
73.0s | ████████████
74.1s | █████████
```

Three silence gaps of 6–8 seconds occur at phase boundaries (segments 11→12, 22→23, 24→25).

## Separate OpenClaw capture sessions confirm consistency

12 WAV segments from earlier (non-debug) sessions show the same pattern — no segment exceeds 7.8 seconds, total captured audio is only 37.6 seconds across all sessions combined.

## Summary

The data suggests that packet loss is not random but correlates with a specific temporal pattern — likely MLS key transition periods. Phase 2 (12–41s after connection) consistently shows degraded capture, with stabilization occurring afterward. This is reproducible across multiple independent sessions.

**Comment by Qjuh** [comment]
AI slop.
